xds/server.go

```diff
@@ -289,23 +426,68 @@ func (l *listenerWrapper) Accept() (net.Conn, error) {
 	if err != nil {
 		return nil, err
 	}
-	return &conn{Conn: c}, nil
+	return &conn{Conn: c, xdsHI: l.xdsHI}, nil
 }
 
 // Close closes the underlying listener. It also cancels the xDS watch
 // registered in Serve() and closes any certificate provider instances created
 // based on security configuration received in the LDS response.
 func (l *listenerWrapper) Close() error {
+	l.closed.Fire()
 	l.Listener.Close()
 	if l.cancelWatch != nil {
 		l.cancelWatch()
 	}
+	if l.cachedIdentity != nil {
+		l.cachedIdentity.Close()
+	}
```

Comment:
I think there's a race between this function and a security update happening.  One way to avoid that would be to access this field atomically; doing a `Load` here and a `Swap` when a new one is assigned due to an update.  Or add a mutex.

Reply:
Thanks. Added a mutex.